### PR TITLE
Helper: First pass at merging the Helper plugin into WooCommerce

### PR DESCRIPTION
**Pull request by kovshenin**
To keep non WordPress.org-hosted extensions up to date, WooCommerce currently uses a separate plugin called WooCommerce Helper. We're working towards bringing all the functionality that plugin provides (and more!) into WooCommerce Core. This first pass allows users to: 

* Connect their store to a WooCommerce.com account
* List their available product subscriptions
* Activate/deactivate subscriptions and extensions
* Receive package updates for active subscriptions
* Receive update notifications for inactive/unavailable subscriptions

If you experience any trouble while testing this, please turn on `WP_DEBUG` and look at the `helper` log in System Status for more insight.

### CODE REVIEW

File: assets/css/helper.scss [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+------------------------------------------------------------------------------*/
+
+.wc-helper {
+	.striped > tbody > :nth-child(odd),
+	ul.striped > :nth-child(odd),
+	.alternate {
+		background-color: #ffffff;
+	}
+
+	table.widefat, .wp-editor-container, .stuffbox, p.popular-tags, .widgets-holder-wrap, .popular-tags, .feature-filter, .imgedit-group, .comment-ays {
```
Comment (mikejolley) [inline assets/css/helper.scss]: Place each selector on a new line (https://make.wordpress.org/core/handbook/best-practices/coding-standards/css/#structure).

File: includes/admin/helper/class-wc-helper-api.php [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+			self::_authenticate( $url, $args );
+		}
+
+		// Disable SSL verification for the dev environment.
+		if ( defined( 'WP_DEBUG' ) && WP_DEBUG && strpos( self::$api_base, 'https://woocommerce.dev/' ) === 0 ) {
+			$args['sslverify'] = false;
+		}
+
+		// TODO: Check response signatures on certain endpoints.
+		return wp_remote_request( $url, $args );
```
Comment (mikejolley) [inline includes/admin/helper/class-wc-helper-api.php]: Question: Default timeout is 3 secs. Is this sufficient?

Comment (mikejolley) [inline includes/admin/helper/class-wc-helper-api.php]: `wp_safe_remote_request` should be used here? `woocommerce_helper_api_base` is filterable.

File: includes/admin/helper/class-wc-helper-plugin-info.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		if ( empty( $args->slug ) ) {
+			return $response;
+		}
+
+		$found_plugin = null;
+
+		// Look through local Woo plugins by slugs.
+		foreach ( WC_Helper::get_local_woo_plugins() as $plugin ) {
+			$slug = dirname( $plugin['_filename'] );
+			if ( dirname( $plugin['_filename'] ) == $args->slug ) {
```
Comment (mikejolley) [inline includes/admin/helper/class-wc-helper-plugin-info.php]: ===

File: includes/admin/helper/views/html-main.php [hunk 4]
```diff
@@ -0,0 +1,188 @@
+<?php defined( 'ABSPATH' ) or exit(); ?>
+
+<div class="wrap wc-helper">
+	<h1><?php _e( 'WooCommerce Helper', 'woocommerce' ); ?></h1>
+
+	<?php include( WC_Helper::get_view_filename( 'html-section-notices.php' ) ); ?>
+	<?php include( WC_Helper::get_view_filename( 'html-section-account.php' ) ); ?>
+
+	<h2>Subscriptions</h2>
```
Comment (mikejolley) [inline includes/admin/helper/views/html-main.php]: This file is missing localisation on all many of the strings - flagging this one but there are more below.

File: includes/admin/helper/views/html-main.php [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+							<a href="#">[?]</a><!-- Sub is active but expired -->
+						<?php endif; ?>
+
+						<?php if ( $update_available && ! $connected ) : ?>
+							<?php echo esc_html( $updates[ $product_id ]['version'] ); ?> available
+							<a href="#">[?]</a><!-- Subscription must be active to receive updates -->
+						<?php endif; ?>
+
+						<?php if ( ! $installed ) : ?>
+							<strong><?php echo $updates[ $product_id ]['version']; ?> available</a></strong>
```
Comment (mikejolley) [inline includes/admin/helper/views/html-main.php]: Missing escape

File: includes/admin/helper/views/html-main.php [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+							<strong><?php echo date( 'F jS, Y', $subscription['expires'] ); ?></strong>
+						<?php elseif ( $subscription['autorenew'] ) : ?>
+							<span class="renews" style="color: black;">Auto renews on:</span><br>
+							<span><?php echo date( 'F jS, Y', $subscription['expires'] ); ?></span>
+						<?php elseif ( $subscription['expiring'] ) : ?>
+							<strong><span class="expiring" style="color: orange;">Expiring soon!</span></strong><br>
+							<strong><?php echo date( 'F jS, Y', $subscription['expires'] ); ?></strong>
+						<?php else : ?>
+							<span class="renews" style="color: black;">Expires on:</span><br>
+							<span><?php echo date( 'F jS, Y', $subscription['expires'] ); ?></span>
```
Comment (mikejolley) [inline includes/admin/helper/views/html-main.php]: `date_i18n` here and lines 84, 81, 78

File: includes/admin/helper/views/html-oauth-start.php [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+<div class="wrap wc-helper">
+	<h1><?php _e( 'WooCommerce Helper', 'woocommerce' ); ?></h1>
+	<?php include( WC_Helper::get_view_filename( 'html-section-notices.php' ) ); ?>
+
+		<div class="start-container">
+			<div class="text">
+				<img src="https://woocommerce.com/wp-content/themes/woomattic/images/logo-woocommerce@2x.png" alt="WooCommerce" style="width:180px;">
+
+				<?php if ( ! empty( $_GET['wc-helper-status'] ) && $_GET['wc-helper-status'] = 'helper-disconnected' ) : ?>
```
Comment (mikejolley) [inline includes/admin/helper/views/html-oauth-start.php]: `$_GET['wc-helper-status'] = 'helper-disconnected'`

you're setting GET to `'helper-disconnected'` here.

File: includes/admin/helper/class-wc-helper-updater.php [hunk 8]
```diff
@@ -0,0 +1,218 @@
+<?php
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+class WC_Helper_Updater {
```
Comment (mikejolley) [inline includes/admin/helper/class-wc-helper-updater.php]: Just a nitpick; some classes and methods throughout the PR are missing docblocks.

**Review by mikejolley — CHANGES_REQUESTED** [review]
Given the code a quick scan and pointed out some potential issues. Leaving additional comments in a sec.

**Comment by mikejolley** [comment]
Had a quick test of this too @kovshenin and functionality worked flawlessly. I have a few expired products on my account which showed fine, as did the plugins I have installed which have no subs.

Just some ideas (forgive me if this has been discussed in other places):

2017-04-24 at 14 47

1. Helper is currently 'always on'. Should we a) add a way to disable this (some users have moaned about the extensions screen before and requested a filter to disable), and b) does it have to always be included, or can be do some kind of detection to see if it's needed or not? i.e. Woo plugin installed.

2. We have 2 sub-menus; extensions and helper. I don't recall where but I do remember a discussion about users expecting 'extensions' to be related to installed extensions. Why can't we merge these sections? Helper would fit quite nicely under there as a sub-menu item or like this:

2017-04-24 at 14 54

The connect banner shown on the helper screen (maybe a more compact version) could also go in there when needed?

cc @mattyza 

3. The actual menu item could show a count or alert symbol if action is required, like the 'plugins' menu does for updates.

File: includes/admin/helper/views/html-oauth-start.php [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+					<p><strong>Sorry to see you go</strong>. Feel free to reconnect again using the button below.</p>
```
Comment (claudiosanches) [inline includes/admin/helper/views/html-oauth-start.php]: This need to be hard coded?

File: includes/admin/helper/views/html-oauth-start.php [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+				<?php endif; ?>
+
+				<h2><?php _e( 'Connect your store to your WooCommerce.com account, and manage your purchases directly from your personal site', 'woocommerce' ); ?></h2>
+				<p><?php _e( 'When you connect your account, your purchases will automatically be listed on your site. You can disconnect the account at anytime. Once your connected, you will receive the updates &amp; support that come with your subscription. ', 'woocommerce' ); ?></p>
+				<p><a class="button-primary" href="<?php echo esc_url( $connect_url ); ?>">Connect your store</a></p>
```
Comment (claudiosanches) [inline includes/admin/helper/views/html-oauth-start.php]: Another hard coded string.

File: includes/admin/helper/views/html-section-account.php [hunk 11]
```diff
@@ -0,0 +1,16 @@
+<?php defined( 'ABSPATH' ) or exit(); ?>
+
+<h2>Your Account</h2>
```
Comment (claudiosanches) [inline includes/admin/helper/views/html-section-account.php]: Other hard coded string.

**Comment by claudiosanches** [comment]
There is a few missing docblocks and hardcoded strings that requires to be fixed.

**Comment by mattyza** [comment]
+1 to merging the "Extensions" and "Helper" screens. I believe this is/was always the intention, going in.

To the point about selectively disabling the Helper portion of the screen, perhaps we could do this automagically, through detection of the products using that screen? If no extensions with product keys are detected, don't show the screen. If extensions with product keys are present, we then show the screen.

Thoughts @mikejolley @kovshenin @garymurray ?

**Comment by garymurray** [comment]
> merging the "Extensions" and "Helper" screens. I believe this is/was always the intention, going in.

Yes, I am going to look at how best we can merge those two screens together in some way. Its next on my agenda.

File: includes/admin/helper/class-wc-helper-api.php (hunk 2 above)
Comment (kovshenin) [inline includes/admin/helper/class-wc-helper-api.php]: Core's default timeout is 5 seconds. Unless WooCommerce changes this with `http_request_timeout` then it should be sufficient.

Comment (kovshenin) [inline includes/admin/helper/class-wc-helper-api.php]: b4ba43f551fe2fb25774455431548911c8637426

File: includes/admin/helper/views/html-oauth-start.php (hunk 7 above)
Comment (kovshenin) [inline includes/admin/helper/views/html-oauth-start.php]: Good catch :)

**Comment by kovshenin** [comment]
Thanks so much for the reviews @claudiosanches and @mikejolley! I think I've addressed all of the issues in the latest chunk of commits.

I also like the idea of merging it with the Extensions screen, we talked about it a bit with @garymurray who's working on some mockups/flows, which will likely require some changes in the views, as well as the "routing" logic in the `WC_Helper` class.

File: includes/admin/helper/class-wc-helper-api.php [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+		return wp_safe_remote_request( $url, $args );
+	}
+
+	/**
+	 * Adds authentication headers to an HTTP request.
+	 *
+	 * @param string $url The request URI.
+	 * @param array $args By-ref, the args that will be passed to wp_remote_request().
+	 *
+	 * @return null
```
Comment (claudiosanches) [inline includes/admin/helper/class-wc-helper-api.php]: This function don't return anything, can just remove it.

File: includes/admin/helper/class-wc-helper-api.php [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+
+		$args['headers'] = array(
+			'Authorization' => 'Bearer ' . $auth['access_token'],
+			'X-Woo-Signature' => $signature,
+		);
+	}
+
+	/**
+	 * Wrapper for self::request().
+	 */
```
Comment (claudiosanches) [inline includes/admin/helper/class-wc-helper-api.php]: Missing docblocks for `$endpoint`, `$args` and returned value.

File: includes/admin/helper/class-wc-helper-api.php [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+	public static function get( $endpoint, $args = array() ) {
+		$args['method'] = 'GET';
+		return self::request( $endpoint, $args );
+	}
+
+	/**
+	 * Wrapper for self::request().
+	 */
```
Comment (claudiosanches) [inline includes/admin/helper/class-wc-helper-api.php]: Missing docblocks for `$endpoint`, `$args` and returned value.

File: includes/admin/helper/class-wc-helper-options.php [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+	exit;
+}
+
+/**
+ * WC_Helper_Options Class
+ *
+ * An interface to the woocommerce_helper_data entry in the wp_options table.
+ */
+class WC_Helper_Options {
+	private static $option_name = 'woocommerce_helper_data';
```
Comment (claudiosanches) [inline includes/admin/helper/class-wc-helper-options.php]: Missing docblock.

File: includes/admin/helper/class-wc-helper-api.php [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+	 * Adds woocommerce.dev to "external" URls if WP_DEBUG is on so
+	 * that wp_http_validate_url allows the request.
+	 *
+	 * @param bool $r Whether the host is external
+	 * @param string $host The requested host.
+	 *
+	 * @return bool True if the host is external, false if local/internal.
+	 */
+	public static function _http_request_host_is_external( $r, $host ) {
+		if ( defined( 'WP_DEBUG' ) && WP_DEBUG && 'woocommerce.dev' === $host ) {
```
Comment (claudiosanches) [inline includes/admin/helper/class-wc-helper-api.php]: This method looks for internal tests only... Not sure if we should include it for all users.

File: includes/admin/helper/class-wc-helper-api.php [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+	 */
+	public static function request( $endpoint, $args = array() ) {
+		$url = self::url( $endpoint );
+
+		if ( ! empty( $args['authenticated'] ) ) {
+			self::_authenticate( $url, $args );
+		}
+
+		// Disable SSL verification for the dev environment.
+		if ( defined( 'WP_DEBUG' ) && WP_DEBUG && strpos( self::$api_base, 'https://woocommerce.dev/' ) === 0 ) {
```
Comment (claudiosanches) [inline includes/admin/helper/class-wc-helper-api.php]: Not sure if we should do it for all users, if we are testing it internally we could create a plugin to do it.

File: includes/admin/helper/class-wc-helper-updater.php [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+			'products' => array(),
+			'errors' => array(),
+		);
+
+		$request = WC_Helper_API::post( 'update-check', array(
+			'body' => json_encode( array( 'products' => $payload ) ),
+			'authenticated' => true,
+		) );
+
+		if ( wp_remote_retrieve_response_code( $request ) != 200 ) {
```
Comment (claudiosanches) [inline includes/admin/helper/class-wc-helper-updater.php]: Needs strict comparison

**Pull request merged**